evaluator/src/boolean/evaluator/fhew.rs

```diff
@@ -0,0 +1,461 @@
+//! Implementation of `BoolEvaluator` using boolean gates in 2020/086 and
+//! blind rotation in 2022/198.
+
+use crate::boolean::evaluator::BoolEvaluator;
+use core::{cell::RefCell, iter::repeat, marker::PhantomData, mem::size_of};
+use itertools::{chain, Itertools};
+use phantom_zone_crypto::{
+    core::{
+        lwe::{
+            self, LweCiphertext, LweCiphertextMutView, LweCiphertextOwned, LweCiphertextView,
+            LweKeySwitchKey, LweKeySwitchKeyOwned, LwePlaintext, LweSecretKeyView,
+        },
+        rgsw::{RgswCiphertext, RgswCiphertextOwned, RgswDecompositionParam},
+        rlwe::{RlweAutoKey, RlweAutoKeyOwned, RlwePlaintext, RlwePlaintextOwned},
+    },
+    scheme::blind_rotation::lmkcdey::{bootstrap, power_g_mod_q},
+    util::{distribution::NoiseDistribution, rng::LweRng},
+};
+use phantom_zone_math::{
+    decomposer::DecompositionParam,
+    modulus::{ElemFrom, Modulus, ModulusOps, NonNativePowerOfTwo},
+    poly::automorphism::AutomorphismMap,
+    ring::RingOps,
+    util::scratch::ScratchOwned,
+};
+use rand::RngCore;
+
+#[derive(Clone, Copy, Debug)]
+pub struct FhewBoolParam {
+    pub modulus: Modulus,
+    pub ring_size: usize,
+    pub auto_decomposition_param: DecompositionParam,
+    pub rgsw_decomposition_param: RgswDecompositionParam,
+    pub lwe_modulus: Modulus,
+    pub lwe_dimension: usize,
+    pub lwe_ks_decomposition_param: DecompositionParam,
+    pub q: usize,
+    pub g: usize,
+    pub w: usize,
+}
+
+impl FhewBoolParam {
+    pub fn embedding_factor(&self) -> usize {
+        2 * self.ring_size / self.q
+    }
+}
+
+#[derive(Clone, Debug)]
+pub struct FhewBoolCiphertext<R: RingOps>(LweCiphertextOwned<R::Elem>, PhantomData<R>);
+
+impl<R: RingOps> FhewBoolCiphertext<R> {
+    pub fn encrypt<'a, T>(
+        ring: &R,
+        sk: impl Into<LweSecretKeyView<'a, T>>,
+        m: bool,
+        noise_distribution: NoiseDistribution,
+        rng: &mut LweRng<impl RngCore, impl RngCore>,
+    ) -> Self
+    where
+        R: ElemFrom<T>,
+        T: 'a + Copy,
+    {
+        let pt = LwePlaintext(NonNativePowerOfTwo::new(2).mod_switch(&(m as u64), ring));
+        let mut ct = LweCiphertext::allocate(ring.ring_size());
+        lwe::sk_encrypt(ring, &mut ct, sk, pt, noise_distribution, rng);
+        Self(ct, PhantomData)
+    }
+
+    pub fn decrypt<'a, T>(&self, ring: &R, sk: impl Into<LweSecretKeyView<'a, T>>) -> bool
+    where
+        R: ElemFrom<T>,
+        T: 'a + Copy,
+    {
+        let pt = lwe::decrypt(ring, sk, &self.0);
+        let m = ring.mod_switch(&pt.0, &NonNativePowerOfTwo::new(2));
+        debug_assert!(m == 0 || m == 1);
+        m == 1
+    }
+}
+
+pub struct FhewBoolEvaluator<R1: RingOps, R2: RingOps> {
```

Comment:
Bootstrapping key is big in size (90 - 100MB). Do we want to the evaluator to own it ? Do we want to have a global reference of the Evaluator. 

Evaluator cannot be shared across threads since scratch must be passed as mutable to functions. 

We can leave this for later, if you think resolving this isn't important at the moment

Reply:
I wanted to differentiate between server key and evaluator. In ideal case, instantiating evaluator must be cause low memory overhead because we may have to do that for each thread given scratch is per thread.

Reply:
Yes, we can simply clone the evaluator for different threads, each evaluator has its own scratch, but they share the same key views.

Reply:
I just bench and find that reusing scratch for bit op is not saving anything, so I remove the `RefCell` and allocate scratch on the fly. This should make the evaluator `Send` and `Sync` I think.